src/tools/io_stress_test.c

```diff
@@ -0,0 +1,912 @@
+                            lfsr_inc_32(lfsr_taps32, &lfsr32));
+                }
+            }
+            current_byte += size_written;
+
+            /* If random wait is enabled, roll for wait. */
+            if (random_mode) {
+                if ((random() % 100) < random_mode) {
+                    usleep(100 * 1000);
+                }
+            }
+        }
+    }
+
+    /* wait for read to finish */
+    pthread_join(read_thread, NULL);
+
+    clock_gettime(CLOCK_MONOTONIC, &end);
+    double diff = (end.tv_sec * 1.0 + end.tv_nsec / 1000000000.0)
+            - (start.tv_sec * 1.0 + start.tv_nsec / 1000000000.0);
+
+    /* final update to the progress bar */
+    pthread_cancel(progressbar_thread);
+    pthread_join(progressbar_thread, NULL);
+    update_progressbar();
+    printf("\n\n");
+
+    if (read_mode == 0) {
+        printf("Sent %zu bytes in %.03f seconds = %.01lf kiB/s\n", current_byte,
+               diff, (current_byte) / diff / 1024);
+    } else {
+        printf("Received %zu bytes in %.03f seconds = %.01lf kiB/s\n",
+               current_byte, diff, (current_byte) / diff / 1024);
+    }
+    printf("Write block size: %zu bytes, read block size: %zu bytes\n",
+           write_block_size, read_block_size);
+    if (use_blocking_functions) {
+        printf("Used blocking function calls (glip_read_b() and glip_write_b())\n");
+    } else {
+        printf("Used non-blocking function calls (glip_read() and glip_write())\n");
+    }
+    glip_close(glip_ctx);
+    return 0;
+}
+
+void* read_from_target(void* ctx_void)
+{
+    struct glip_ctx *ctx = ctx_void;
+
+    uint8_t *read_data = calloc(read_block_size, sizeof(uint8_t));
+    uint16_t *read_data_16 = (uint16_t*) read_data;
+    uint32_t *read_data_32 = (uint32_t*) read_data;
+    assert(read_data);
+
+    size_t size_read;
+
+    /* Used for read mode to generate pseudo random values locally. */
+    uint16_t *rnd_word_16 = calloc(read_block_size, sizeof(uint8_t));
+    uint32_t *rnd_word_32 = (uint32_t*) rnd_word_16;
+    assert(rnd_word_16);
+
+    int rv;
+    while (1) {
+        /* Read data from GLIP device. */
+        if (use_blocking_functions) {
+            rv = glip_read_b(ctx, 0, read_block_size, read_data, &size_read,
+            READ_TIMEOUT_MS);
+        } else {
+            rv = glip_read(ctx, 0, read_block_size, read_data, &size_read);
+        }
+        if (rv != 0 && rv != -ETIMEDOUT) {
+            fprintf(stderr, "Error while reading from GLIP. rv = %d\n", rv);
+            exit_measurement(1);
+        }
+
+        /*
+         * Write stress test:
+         * Only messaged received in write mode are error messages or success
+         * message at end of test.
+         */
+        if (read_mode == 0) {
+            if (size_read > 0) {
+                /* 16-bit interface */
+                if (fifo_width_bytes == 2) {
+                    /* Check for success message. */
+                    if (size_read == 2) {
```

Comment:
Please try to reduce if/else indentation levels, as well as the lines of code in this function, by two measures:
- Return early if possible (see below)
- Create suitably-named small helper functions for the individual subtasks. The Linux kernel guidelines put this quite nicely "Functions should be short and sweet, and do just one thing. They should fit on one or two screenfuls of text (the ISO/ANSI screen size is 80x24, as we all know), and do one thing and do that well."